File: src/commands/addKey.ts
import { ChatInputCommandInteraction, SlashCommandBuilder } from 'discord.js'
import { adminCommand } from '../helpers/commandPermissions'
import { database } from '../helpers/database'

export = {
  data: new SlashCommandBuilder()
    .setName('addkey')
    .setDescription('Adds a Playne key to the database.')
    .addStringOption(option =>
      option.setName('key')
        .setDescription('The key to add.')
        .setRequired(true))
    .setDefaultMemberPermissions(adminCommand())
    .setDMPermission(false),
  async execute (interaction: ChatInputCommandInteraction) {
    const key: string = interaction.options.getString('key') ?? ''

    const keyExists = await database.steamKeys.findFirst({
      where: {
        key
      }
    })

    if (keyExists != null) {
      return interaction.reply({ content: ':x: Key already exists.', ephemeral: true })
    }

    await database.steamKeys.create({
      data: {
        key,
        used: false
      }
    })

    return interaction.reply({ content: ':white_check_mark: Key has been added.', ephemeral: true })
  }
}

File: src/commands/add_key.rs
use crate::commands::{commit_and_say, MessageType};
use crate::database::DatabaseHandler;
use crate::Context;
use anyhow::Result;

/// Adds a Playne key to the database.
#[poise::command(
  slash_command,
  required_permissions = "ADMINISTRATOR",
  rename = "addkey",
  guild_only
)]
pub async fn add_key(
  ctx: Context<'_>,
  #[description = "The Playne key to add"] key: String,
) -> Result<()> {
  let data = ctx.data();

  // We unwrap here, because we know that the command is guild-only.
  let guild_id = ctx.guild_id().unwrap();

  let mut transaction = data.db.start_transaction_with_retry(5).await?;
  if DatabaseHandler::steam_key_exists(&mut transaction, &guild_id, key.as_str()).await? {
    ctx
      .send(|f| f.content(":x: Key already exists.").ephemeral(true))
      .await?;
    return Ok(());
  }

  DatabaseHandler::add_steam_key(&mut transaction, &guild_id, key.as_str()).await?;

  commit_and_say(
    ctx,
    transaction,
    MessageType::TextOnly(":white_check_mark: Key has been added.".to_string()),
    true,
  )
  .await?;

  Ok(())
}
